### PR TITLE
(not complete) splink

### PR DESCRIPTION
**Pull request by npashilkar**
initial code for splink record linkage function for the persons table

### CODE REVIEW

**Comment by npashilkar** [comment]
i moved my splink code from the notebook i was working in to a .py file. however, i was still getting an error in the notebook that said the model wasn't being fully trained because there was not enough data to train all comparisons. it required at minimum hundreds and ideally thousands of rows. i think someone else is also working on a test file for the other deduplication function so maybe i can take a look at that?

File: utils/linkage.py [hunk 1]
```diff
@@ -459,3 +467,26 @@ def get_address_number_from_address_line_1(address_line_1: str) -> str:
...
+
+def splink_dedupe(df, settings, blocking):
+
+      """Given a dataframe, configuration settings, and blocking rules return a deduplicated dataframe
+    Args:
+        df: dataframe with potential duplicates
+        settings: model settings based on splink documentation
+        blocking: list of columns to block on
+    Returns:
+        dataframe with matched ids of matching rows"""
```
Comment (trevorspreadbury) [inline utils/linkage.py]: the spacing on this is strange. Should be no blank line after function signature (`def ..... ):`, and there should be a blank line before args

File: utils/linkage.py [hunk 2]
```diff
@@ -459,3 +467,26 @@ def get_address_number_from_address_line_1(address_line_1: str) -> str:
         elif address_line_1_components[i][1] == "USPSBoxID":
             return address_line_1_components[i][0]
     raise ValueError("Can not find Address Number")
+
+
+def splink_dedupe(df, settings, blocking):
+
+      """Given a dataframe, configuration settings, and blocking rules return a deduplicated dataframe
+    Args:
+        df: dataframe with potential duplicates
```
Comment (trevorspreadbury) [inline utils/linkage.py]: Does the df have any required columns? Does this work for both individuals and organizations?

File: utils/linkage.py [hunk 3]
```diff
@@ -459,3 +467,26 @@ def get_address_number_from_address_line_1(address_line_1: str) -> str:
         elif address_line_1_components[i][1] == "USPSBoxID":
             return address_line_1_components[i][0]
     raise ValueError("Can not find Address Number")
+
+
+def splink_dedupe(df, settings, blocking):
```
Comment (averyschoen) [inline utils/linkage.py]: let's add type notation to the inputs and expected outputs

**Review by averyschoen — CHANGES_REQUESTED** [review]
add your tests to the test_linkage.py file

File: utils/linkage.py [hunk 4]
```diff
@@ -635,3 +636,39 @@ def get_address_number_from_address_line_1(address_line_1: str) -> str:
         elif address_line_1_components[i][1] == "USPSBoxID":
             return address_line_1_components[i][0]
     raise ValueError("Can not find Address Number")
+
+
+def splink_dedupe(
+    df: pd.DataFrame, settings: dict, blocking: list
+) -> pd.DataFrame:
+    """Given the individuals or organizations dataframes, the corresponding
```
Comment (averyschoen) [inline utils/linkage.py]: Given a dataframe and config, dedupe with splink

File: utils/linkage.py [hunk 5]
```diff
@@ -635,3 +636,39 @@ def get_address_number_from_address_line_1(address_line_1: str) -> str:
...
+    configuration settings, and corresponding blocking rules return a
+    deduplicated dataframe
+
+    Configuration settings and blocking can be found in constants.py as
+    individuals_settings, indivduals_blocking, organizations_settings,
+    organizations_blocking
+
+    Args:
+        df: individuals or organizations table
```
Comment (averyschoen) [inline utils/linkage.py]: generalize

File: utils/linkage.py [hunk 6]
```diff
@@ -635,3 +636,39 @@ def get_address_number_from_address_line_1(address_line_1: str) -> str:
...
+        settings: individuald or configuration settings
+            (based on splink documentation)
+        blocking: list of columns to block on for the table
+            (cuts dataframe into parts based on above blocks )
+    Returns:
+        dataframe with matched ids of matching rows
```
Comment (averyschoen) [inline utils/linkage.py]: more detail on the format of this

**Review by averyschoen — CHANGES_REQUESTED** [review]

File: utils/linkage.py [hunk 7]
```diff
@@ -635,3 +636,61 @@ def get_address_number_from_address_line_1(address_line_1: str) -> str:
         elif address_line_1_components[i][1] == "USPSBoxID":
             return address_line_1_components[i][0]
     raise ValueError("Can not find Address Number")
+
+
+def splink_dedupe(
+    df: pd.DataFrame, settings: dict, blocking: list
+) -> pd.DataFrame:
+    """Given a dataframes, the corresponding
```
Comment (averyschoen) [inline utils/linkage.py]: shorten this line

File: notebooks/splink.ipynb
Comment (averyschoen) [inline notebooks/splink.ipynb:1]: remove this

**Review by averyschoen — CHANGES_REQUESTED** [review]

**Review by averyschoen — CHANGES_REQUESTED** [review]
update to call convert_to_dict(), delete write to csv step, return deduplicated csv with the cols in the schema -- i.e. inputted columns should be what the returned columns are

**Review by averyschoen — APPROVED** [review]

**Pull request merged**